### PR TITLE
feat(layer2): add Aurora Optimistic Network

### PR DESCRIPTION
**Pull request by developerfred**
### Aurora Layer 2
https://doc.aurora.dev/develop/compat/evm https://github.com/aurora-is-near

### Aurora Dex

- Dodoex ✅ 
- WannaSwap ✅ 
- Trisolaris ✅  https://github.com/dimensiondev/maskbook/issues/5818

### Token List
 -  [x] https://github.com/DimensionDev/Mask-Token-List/pull/46 ✅

### Contracts  Mainnet 
- [x] BalanceChecker 
- [x] MultiCall

### Contracts  Testnet
- [x] BalanceChecker  ✅ 
- [x] MultiCall ✅ 
- [x] Maskbook Contracts  ✅ 

### Sourcify
- [ ] https://github.com/ethereum/sourcify/pull/621

### Aurora Bridge
https://rainbowbridge.app/

### Faucet
https://help.aurora.dev/article/22-where-can-i-get-tokens-for-the-testnet

### Setup Metadata

- [x] Add a logo image
  - `packages/mask/src/plugins/EVM/assets/`
- [x] Add metadata data
  - `packages/web3-shared/evm/assets/chains.json`
- [x] Add an HTTP RPC node
  - `packages/web3-constants/evm/rpc.json`
- [x] Set network descriptor
  - `PLUGIN_NETWORKS` in `packages/mask/src/plugins/EVM/constants.ts`

### Setup External APIs

| Provider      | Configuration Path                               | status |
| ------------- | ------------------------------------------------ |-------|
| DeBank        | `packages/web3-constants/evm/debank.json`        | ✅ |
| CoinGecko     | `packages/web3-constants/evm/coingecko.json`     |✅ |
| CoinMarketCap | `packages/web3-constants/evm/coinmarketcap.json` |✅ 
| Zerion        | `packages/web3-constants/evm/zerion.json`        | ✅ |

| DEX       | Pull Request Link                                  |status |
| --------- | -------------------------------------------------- |-------|
| DODO      | https://github.com/DimensionDev/Maskbook/pull/5323/ |✅ |
| WannaSwap |https://github.com/DimensionDev/Maskbook/pull/5323/ |✅ |
| Trisolaris |https://github.com/DimensionDev/Maskbook/pull/5323/ |✅ |

## Testing Checklist

Before we ship the chain to the user, we need to do basic ability checks.

- [x] Check if the asset list and transaction history on the dashboard page work when you choose the chain as the network.
  - Setup the chain for DeBank API.
- [x] Check if the trending view is working. Try to hover a new chain token in any tweets.
  - Setup the chain for Coingecko API.
  - Setup the chain for CoinMarketCap API.

image

- [x] Check if the gas estimate dialog is working.
  - Setup the chain for CoinGecko API.

Goto `chrome-extension://jkoeaghipilijlahjplgbfiocjhldnap/dashboard.html#/wallets/transfer` and check the estimated USD value is working.

image

- [x] Trade with the DEX on the chain.

  - Integrate a DEX for the chain.

- [x] Send a transaction and check if the explorer links are working.
  - Setup metadata and RPC for the chain.

### CODE REVIEW

**Review by Jack-Works — DISMISSED** [review]

**Comment by septs** [comment]
plz resolve conflict.
CleanShot 2021-12-30 at 08 09 24@2x

**Comment by developerfred** [comment]
@septs All conflicts have been resolved.  

- [List codingsh Pull Requests ](https://github.com/DimensionDev/Maskbook/pulls?q=is%3Aopen+is%3Apr+author%3Adeveloperfred+sort%3Aupdated-desc+)

**Review by Jack-Works — DISMISSED** [review]

File: packages/encryption/src/index.ts [hunk 1]
```diff
@@ -19,12 +19,12 @@ export {
     ErrorReasons,
     type DecryptOptions,
     type DecryptIO,
-    type DecryptEphemeralECDH_PostKey,
```
Comment (Jack-Works) [inline packages/encryption/src/index.ts]: hi please revert changes in the packages/encryption folder thanks

File: packages/mask/src/plugins/Trader/trader/useGetTradeContext.ts [hunk 2]
```diff
@@ -13,23 +13,28 @@ import {
...
+    WANNASWAP_CUSTOM_BASES,
+    TRISOLARIS_BASE_AGAINST_TOKENS,
+    TRISOLARIS_CUSTOM_BASES,
 } from '../constants'
 import { unreachable } from '@dimensiondev/kit'
 import { TargetChainIdContext } from './useTargetChainIdContext'
 
 export function useGetTradeContext(tradeProvider?: TradeProvider) {
     const { targetChainId: chainId } = TargetChainIdContext.useContainer()
+    const DEX_TRADE = getTraderConstants(chainId)
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/trader/useGetTradeContext.ts]: Nice change 👍 

```ts
const DEX_TRADE = getTraderConstants(chainId)
```

Moving this line inside `useMemo` following would make the code compact, also benefit from memo.

File: packages/mask/src/plugins/Trader/trader/useTradeContext.ts [hunk 3]
```diff
@@ -21,16 +25,17 @@ export const TradeContext = createContext<TradeContext_ | null>(null)
 
 export function useTradeContext(tradeProvider: TradeProvider) {
     const { targetChainId: chainId } = TargetChainIdContext.useContainer()
+    const DEX_TRADE = getTraderConstants(chainId)
```
Comment (UncleBill) [inline packages/mask/src/plugins/Trader/trader/useTradeContext.ts]: dito

File: packages/mask/src/plugins/Trader/trader/useGetTradeContext.ts (hunk 2 above)
Comment (developerfred) [reply packages/mask/src/plugins/Trader/trader/useGetTradeContext.ts]: Ok I will change

Comment (septs) [reply packages/mask/src/plugins/Trader/trader/useGetTradeContext.ts]: @UncleBill use `useTraderConstants` replace it ?

**Comment by guanbinrui** [comment]
Hi @developerfred, plz follow the [evm-integration-doc](https://github.com/DimensionDev/Maskbook/blob/develop/docs/evm-integration.md) to ensure nothing left.

**Comment by developerfred** [comment]
> Hi @developerfred, plz follow the [evm-integration-doc](https://github.com/DimensionDev/Maskbook/blob/develop/docs/evm-integration.md?rgh-link-date=2022-01-11T07%3A47%3A00Z) to ensure nothing left.


need 
- [x] Other contracts from the Mask team ( Mainnet)

**Comment by guanbinrui** [comment]
> > Hi @developerfred, plz follow the [evm-integration-doc](https://github.com/DimensionDev/Maskbook/blob/develop/docs/evm-integration.md?rgh-link-date=2022-01-11T07%3A47%3A00Z) to ensure nothing left.
> 
> need
> 
> * [ ]  Other contracts from the Mask team ( Mainnet)

Thanks! @UncleBill will help to deploy those contracts.

**Comment by developerfred** [comment]
> Thanks! @UncleBill will help to deploy those contracts.

Thanks! @UncleBill What would be the help, will you implement the contracts at Aurora? or can I continue the implementation?

**Comment by developerfred** [comment]
@guanbinrui  and @UncleBill the deployment has already been done.

**Comment by UncleBill** [comment]
@developerfred I have asked our contract developer to deploy contracts to the networks that being integrated.

**Comment by UncleBill** [comment]
@developerfred  contracts have been deployed on Aurora, you can check out the addresses here at https://github.com/DimensionDev/misc_smart_contract/pull/3

**Comment by developerfred** [comment]
> @developerfred contracts have been deployed on Aurora, you can check out the addresses here at https://github.com/DimensionDev/misc_smart_contract/pull/3

I don't have access to this resource, I had already deployed it here:

https://github.com/DimensionDev/Maskbook/pull/5323/commits/59463962219be88841f53b906160ee8f91f45973

**Comment by UncleBill** [comment]
Here they are. @developerfred 

### BalanceChecker

| Chain            | Address                                  |
| ---------------- | ---------------------------------------- |
| Aurora-testnet   | [0x57E2AAB7][BalanceChecker-aurora-test] |
| Aurora           | [0xC119574D][BalanceChecker-aurora]      |

[BalanceChecker-aurora-test]: https://explorer.testnet.aurora.dev/address/0x57E2AAB712E9c61CA55A6402223DbEe3d4eE09aa/transactions
[BalanceChecker-aurora]: https://explorer.mainnet.aurora.dev/address/0xC119574D5Fb333F5AC018658D4d8b5035E16bf39/transactions

### UniswapInterfaceMulticall

| Chain            | Address                                                                                  |
| ---------------- | ---------------------------------------------------------------------------------------- |
| Aurora-testnet   | [0x7e0Ac0A9][BalanceChecker-aurora-test]                                                 |
| Aurora           | [0x6cc1b105][BalanceChecker-aurora]                                                      |

[BalanceChecker-aurora-test]: https://explorer.testnet.aurora.dev/address/0x7e0Ac0A9341017a687f8d8bD4287104Ea86c5924/transactions
[BalanceChecker-aurora]: https://explorer.mainnet.aurora.dev/address/0x6cc1b1058F9153358278C35E0b2D382f1585854B/transactions

### ~~Multicall~~


| Chain            | Address                                                                                                           |
| ---------------- | ----------------------------------------------------------------------------------------------------------------- |
| Aurora-test      | [0x4A77E797](https://explorer.testnet.aurora.dev/address/0x4A77E797031257db72F7D2C3Ec08a4FAc5c8CfE9/transactions) |
| Aurora           | [0x71834a3F](https://explorer.mainnet.aurora.dev/address/0x71834a3FDeA3E70F14a93ED85c6be70925D0CAd9/transactions) |

**Comment by developerfred** [comment]
@UncleBill  will you use the contracts with the code without even sourcify?

**Comment by UncleBill** [comment]
> @UncleBill will you use the contracts with the code without even sourcify?

They are developed and deployed by our contract developer. We could trust them.

**Comment by UncleBill** [comment]
"Wrong network" shows after switching to Aurora

image


image

You should make sure `getNetworkTypeFromChainId` inside has a same name to `CHAINS` constants configuration.


```diff

diff --git a/packages/web3-shared/evm/assets/chains.json b/packages/web3-shared/evm/assets/chains.json
index 02d2a9cb8..1c431d184 100644
--- a/packages/web3-shared/evm/assets/chains.json
+++ b/packages/web3-shared/evm/assets/chains.json
@@ -2604,7 +2604,7 @@
     },
     {
         "name": "Aurora MainNet",
-        "chain": "NEAR",
+        "chain": "Aurora",
         "network": "mainnet",
         "rpc": ["https://rpc.mainnet.aurora.dev:8545"],
         "faucets": [],
```


Better to go through following operations to make sure all configuration are correct:


- [ ] Trade with the DEX on the chain.
- [ ] Send a transaction and check if the explorer links are working.
  - Setup metadata and RPC for the chain.

**Comment by developerfred** [comment]
You should make sure getNetworkTypeFromChainId inside has a same name to CHAINS constants configuration.
done ✅

File: packages/web3-constants/evm/token.json [hunk 4]
```diff
@@ -13,7 +13,10 @@
         "Arbitrum_Rinkeby": "0xB47e6A5f8b33b3F17603C83a0535A9dcD7E32681",
         "xDai": "0xe91D153E0b41518A2Ce8Dd3D7944Fa863463a97d",
         "Celo": "0x471EcE3750Da237f93B8E339c536989b8978a438",
-        "Fantom": "0x21be370d5312f44cb42ce377bc9b8a0cef1a4c83"
+        "Fantom": "0x21be370d5312f44cb42ce377bc9b8a0cef1a4c83",
+        "Aurora": "0xC9BdeEd33CD01541e1eeD10f90519d2C06Fe3feB",
+        "Aurora_Testnet": "",
+        "Aurora_Betanet": ""
```
Comment (UncleBill) [inline packages/web3-constants/evm/token.json]: I don't see Aurora_Betanet around other files, do we need it here?

**Comment by UncleBill** [comment]
To make a DEX work, we should config common token in `packages/web3-constants/evm/token.json` (Hope this tip could help you fix other PRs too. Plus, I will update it in our docs later)



```diff
diff --git a/packages/mask/src/plugins/Trader/constants/trisolaris.ts b/packages/mask/src/plugins/Trader/constants/trisolaris.ts
index f443728c5..cee359af1 100644
--- a/packages/mask/src/plugins/Trader/constants/trisolaris.ts
+++ b/packages/mask/src/plugins/Trader/constants/trisolaris.ts
@@ -1,5 +1,5 @@
 import { ChainId } from '@masknet/web3-shared-evm'
-import { DAI, MSKA, MSKB, MSKC, USDC, USDT, fUSDT, WBTC, WNATIVE, WNATIVE_ONLY } from './trader'
+import { DAI, MSKA, MSKB, MSKC, USDC, USDT, WBTC, WNATIVE, WNATIVE_ONLY } from './trader'
 import type { ERC20AgainstToken, ERC20TokenCustomizedBase } from './types'
 
 /**
@@ -12,5 +12,5 @@ export const TRISOLARIS_BASE_AGAINST_TOKENS: ERC20AgainstToken = {
     ...WNATIVE_ONLY,
     [ChainId.Rinkeby]: [WNATIVE, MSKA, MSKB, MSKC].map((x) => x[ChainId.Rinkeby]),
     [ChainId.Matic]: [WNATIVE, USDC, WBTC, DAI, USDT].map((x) => x[ChainId.Matic]),
-    [ChainId.Aurora]: [WNATIVE, DAI, USDT, USDC, fUSDT, WBTC].map((x) => x[ChainId.Aurora]),
+    [ChainId.Aurora]: [WNATIVE, DAI, USDT, USDC, WBTC].map((x) => x[ChainId.Aurora]),
 }
diff --git a/packages/web3-constants/evm/token-list.json b/packages/web3-constants/evm/token-list.json
index 6b78afb14..474b6dc5d 100644
--- a/packages/web3-constants/evm/token-list.json
+++ b/packages/web3-constants/evm/token-list.json
@@ -27,9 +27,7 @@
         "xDai": ["https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/100/tokens.json"],
         "Celo": ["https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/42220/tokens.json"],
         "Fantom": ["https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/250/tokens.json"],
-        "Aurora": [
-            "https://raw.githubusercontent.com/DimensionDev/Mask-Token-List/gh-pages/latest/1313161554/tokens.json"
-        ],
+        "Aurora": ["https://raw.githubusercontent.com/trisolaris-labs/tokens/master/lists/1313161554/list.json"],
         "Aurora_Testnet": "",
         "Aurora_Betanet": ""
     }
diff --git a/packages/web3-constants/evm/token.json b/packages/web3-constants/evm/token.json
index c5e1aa679..f2a11cd30 100644
--- a/packages/web3-constants/evm/token.json
+++ b/packages/web3-constants/evm/token.json
@@ -33,7 +33,7 @@
         "xDai": "0xDDAfbb505ad214D7b80b1f830fcCc89B60fb7A83",
         "Celo": "0x2A3684e9Dc20B857375EA04235F2F7edBe818FA7",
         "Fantom": "0x04068da6c83afcfa0e13ba15a6696662335d5b75",
-        "Aurora": "",
+        "Aurora": "0xb12bfca5a55806aaf64e99521918a4bf0fc40802",
         "Aurora_Testnet": "",
         "Aurora_Betanet": ""
     },
@@ -52,7 +52,7 @@
         "xDai": "0x4ECaBa5870353805a9F068101A40E0f32ed605C6",
         "Celo": "0xb020d981420744f6b0fedd22bb67cd37ce18a1d5",
         "Fantom": "",
-        "Aurora": "",
+        "Aurora": "0x4988a896b1227218e4a686fde5eabdcabd91571f",
         "Aurora_Testnet": "",
         "Aurora_Betanet": ""
     },
@@ -280,7 +280,7 @@
         "xDai": "",
         "Celo": "",
         "Fantom": "0x8D11eC38a3EB5E956B052f67Da8Bdc9bef8Abf3E",
-        "Aurora": "",
+        "Aurora": "0xe3520349f477a5f6eb06107066048508498a291b",
         "Aurora_Testnet": "",
         "Aurora_Betanet": ""
     },
@@ -508,7 +508,7 @@
         "xDai": "0x8e5bBbb09Ed1ebdE8674Cda39A0c169401db4252",
         "Celo": "0xBe50a3013A1c94768A1ABb78c3cB79AB28fc1aCE",
         "Fantom": "0x321162Cd933E2Be498Cd2267a90534A804051b11",
-        "Aurora": "",
+        "Aurora": "0xf4eb217ba2454613b15dbdea6e5f22276410e89e",
         "Aurora_Testnet": "",
         "Aurora_Betanet": ""
     },
```

File: packages/web3-constants/evm/token.json (hunk 4 above)
Comment (developerfred) [reply packages/web3-constants/evm/token.json]: I just removed the beta network.

**Comment by UncleBill** [comment]
@developerfred Could you add me to your forked repo? I wanna push some changes to your branch.

**Comment by developerfred** [comment]
Screenshot 2022-01-14 at 19 46 56
Screenshot 2022-01-14 at 19 47 42

**Comment by UncleBill** [comment]
I change some code, now the swap works.

**Comment by UncleBill** [comment]
@DimensionDev/qa Please help to verify.

**Comment by developerfred** [comment]
> I change some code, now the swap works.

Looks like my last pull request commit is gone, swap is already working.  but thank you for improving

**Comment by developerfred** [comment]
> > I change some code, now the swap works.
> 
> Looks like my last pull request commit is gone, swap is already working. but thank you for improving

Screenshot 2022-01-15 at 11 42 13

**Comment by UncleBill** [comment]
> > > I change some code, now the swap works.
> > 
> > 
> > Looks like my last pull request commit is gone, swap is already working. but thank you for improving
> 
> Screenshot 2022-01-15 at 11 42 13

Do you forget to push? I don't see this commit in branch `feat/AuroraNearEVM`

image


https://github.com/developerfred/Maskbook/commit/be4432a6 is 404 right now.

**Comment by developerfred** [comment]
> > > > I change some code, now the swap works.
> > > 
> > > 
> > > Looks like my last pull request commit is gone, swap is already working. but thank you for improving
> > 
> > 
> > Screenshot 2022-01-15 at 11 42 13
> 
> Do you forget to push? I don't see this commit in branch `feat/AuroraNearEVM`
> 
> image
> 
> [developerfred@be4432a](https://github.com/developerfred/Maskbook/commit/be4432a6) is 404 right now.

ah ok, no worries!
efb1a10cac95b4d122013088802075efeadce2f0

File: packages/mask/src/plugins/Trader/trader/openocean/constants.ts [hunk 5]
```diff
@@ -0,0 +1 @@
+export const OPENOCEAN_SUPPORTED_CHAINS = [1, 56, 100, 137, 250, 42161, 43114]
```
Comment (guanbinrui) [inline packages/mask/src/plugins/Trader/trader/openocean/constants.ts]: Use `ChainId`.

**Comment by developerfred** [comment]
🎉

**Pull request merged**